metals/src/main/scala/scala/meta/internal/metals/LSPFileSystemProvider.scala

```diff
@@ -0,0 +1,249 @@
+package scala.meta.internal.metals
+
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.stream.Collectors
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.clients.language.MetalsLanguageClient
+
+final case class FSReadDirectoryResponse(
+    name: String,
+    isFile: Boolean,
+)
+final case class FSReadDirectoriesResponse(
+    name: String,
+    directories: Array[FSReadDirectoryResponse],
+    error: String,
+)
+
+final case class FSReadFileResponse(
+    name: String,
+    value: String,
+    error: String,
+)
+
+final case class FSStatResponse(
+    name: String,
+    isFile: Boolean,
+    error: String,
+)
+
+final class LSPFileSystemProvider(
+    languageClient: MetalsLanguageClient,
+    uriMapper: URIMapper,
+    fileDecoderProvider: FileDecoderProvider,
+    clientConfig: ClientConfiguration,
+)(implicit ec: ExecutionContext) {
+
+  // cache the last few decompiled class files as they're repeatedly accessed by VSCode once opened or referenced
+  private val cache = LRUCache[Path, String](10)
```

Comment:
Do we need it? Most likely the user will have the class open until it's not needed.

Reply:
The problem here is that there are a lot of LSP messages flying around when using a document and the message handlers nearly always require the text of the document.
The `metals/didFocusTextDocument` message - which appears to be a Metals specific message? - doesn't pass the text of the document (unlike the standard LSP messages: `textDocument/didOpen`, `textDocument/didChange`, `textDocument/didClose`) and so every time there's a change of focus involving a decompiled class file, Metals has to decompile it again as it needs the text.  This takes a few seconds, especially as it's an external process and we don't use CFR as a library which would be much faster.
Ideally we'd change `metals/didFocusTextDocument` to include the text and pass that text into `interactiveSemanticdbs.didFocus(path)`, `syntheticsDecorator.publishSynthetics(path)`, `worksheetProvider.onDidFocus(path)` which all appear to reload the text from disk.  There's a lot of fileIO going on for all of these.

Reply:
Ok sure, makes sense to use it then.